src/workerd/io/worker.c++

```diff
@@ -2303,6 +2305,39 @@ public:
     }
   }
 
+  kj::Promise<void> dispatchProtocolMessages(kj::MutexGuarded<MessageQueue>& incomingQueue) {
+    auto i = kj::atomicAddRef(*this->state.lockExclusive()->get()->isolate);
+    auto asyncLock = co_await i->takeAsyncLockWithoutRequest(nullptr);
+
+    auto lockedState = this->state.lockExclusive();
+    v8_inspector::V8InspectorSession& session = *lockedState->get()->session;
+    Isolate& isolate = const_cast<Isolate&>(*lockedState->get()->isolate);
+    jsg::V8StackScope stackScope;
+    Isolate::Impl::Lock recordedLock(isolate, asyncLock, stackScope);
+    for (;;) {
+      kj::String message;
+      {
+        auto lockedMessageQueue = incomingQueue.lockExclusive();
+        if (lockedMessageQueue->status != MessageQueue::Status::ACTIVE) {
+          co_return;
+        }
+
+        if (lockedMessageQueue->head == lockedMessageQueue->messages.size()) {
+          co_return;
+        }
+
+        message = kj::mv(lockedMessageQueue->messages[lockedMessageQueue->head++]);
+        if (lockedMessageQueue->head == lockedMessageQueue->messages.size()) {
```

Comment:
If the queue only drops old items when empty, could this lead to a situation where a producer is running just slightly faster than a consumer, but the memory usage grows out of proportion to the number of pending items? (It'd be linear in (processed + pending) items)

(Meta-question: and is this situation something that's probable enough for us to care?)

Reply:
This is a great point in general.

In this specific case, this code only runs on the I/O thread that sends and receives messages. The I/O thread has 3 co-routines: one that receives messages from a debugger over a websocket, one that dispatches the received debugger messages, and one that sends debugger responses from V8 to the outside agent. Whilst dispatching messages here, the I/O thread is not adding messages to the queue.

When debugging a breakpoint (or debugger bytecode), the isolate is locked and V8 runs in a 'pause loop' so the I/O thread can't dispatch messages because it can't get the isolate lock. This means the I/O thread just has the send message co-routine and the receive message co-routine as candidates to run. In this situation, the isolate thread takes all the messages in the queue and dispatches them.

Reply:
Actually, I may have spoken too early  in the latter paragraph...looking into it. ☺️

Reply:
Okay, the code I was looking at is dead and removed in the latest push:
 https://github.com/cloudflare/workerd/compare/b84fee5172c169be119efbedc1bd2c792b39310c..ae0a58fee3adde0b329d838b8b78a50a4727e19e

There is scope to do things better there, and I'll have a go at getting that done later this week.